public/app/core/logs_model.ts

```diff
@@ -619,30 +619,66 @@ function aggregateFields(dataFrames: DataFrame[], config: FieldConfig): DataFram
 
 const LOGS_VOLUME_QUERY_DEFAULT_TIMEOUT = 60000;
 
+/**
+ * Maps formatted interval to moment.js duration unit so it can be used with startOf to round buckets start/end time
+ */
+export const LOG_VOLUME_BUCKETS: Record<string, DurationUnit> = {
+  '1ms': 'millisecond',
+  '1s': 'second',
+  '1m': 'minute',
+  '1h': 'hour',
+  '1d': 'day',
+};
+
 type LogsVolumeQueryOptions<T extends DataQuery> = {
   timeout?: number;
   extractLevel: (dataFrame: DataFrame) => LogLevel;
-  targets: T[];
-  range: TimeRange;
 };
 
-/**
- * Creates an observable, which makes requests to get logs volume and aggregates results.
- */
-export function queryLogsVolume<T extends DataQuery>(
+export type LogVolumeMeta = {
+  cacheInfo: {
+    targets: DataQuery[];
+    absoluteRange: AbsoluteTimeRange;
+  };
+  bucketSize: number;
+};
+
+export const hasLogsVolumeMeta = (custom: any): custom is LogVolumeMeta => {
+  return (
+    custom !== undefined &&
+    custom.cacheInfo !== undefined &&
+    typeof custom.cacheInfo.targets === 'object' &&
```

Comment:
Since this can throw if `custom` doesn't have property `cacheInfo`, should the function body be wrapped in a `try` block?

Reply:
Good point. There's a check for `custom.cacheInfo !== undefined` but could potentially throw if `cacheInfo` was `null` 🤔 Might be better to wrap it try catch 👍

Reply:
I've added `custom !== null && typeof custom === 'object'` to better check for nulls in 52ba83be94c577be3ea99dc98717d74668d2a04a

Reply:
Wouldn't `typeof custom?.cacheInfo?.targets === ...` work here and make this shorter?

Reply:
Good point! Updated in 5b7a9d26ed6f47df8835f318868f5d762b18656d